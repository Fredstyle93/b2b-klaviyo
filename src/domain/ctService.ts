import { AuthMiddlewareOptions, Client, ClientBuilder, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import { ApiRoot, Payment, createApiBuilderFromCtpClient, Customer, Order } from '@commercetools/platform-sdk'
import logger from '../utils/log';
import fetch from 'node-fetch';
import { StatusError } from '../types/errors/StatusError';

type ByProjectKeyRequestBuilder = ReturnType<ApiRoot['withProjectKey']>;

let ctClient: Client;

export const createCtClient = () => {
    if (!ctClient) {
        const ctApiCredentials = JSON.parse(process.env.CT_API_CLIENT || '{"clientId":"","secret":""}');

        const authMiddlewareOptions: AuthMiddlewareOptions = {
            host: process.env.CT_AUTH_URL || '',
            projectKey: process.env.CT_PROJECT_ID || '',
            credentials: {
                clientId: ctApiCredentials.clientId || '',
                clientSecret: ctApiCredentials.secret || '',
            },
            scopes: [process.env.CT_SCOPES || ''],
            fetch,
        };

        const httpMiddlewareOptions: HttpMiddlewareOptions = {
            host: process.env.CT_API_URL || '',
            fetch,
        };

        ctClient = new ClientBuilder()
            .withClientCredentialsFlow(authMiddlewareOptions)
            .withHttpMiddleware(httpMiddlewareOptions)
            // .withLoggerMiddleware()
            .build();
    }
    return ctClient;
};

const getApiRoot = (): ByProjectKeyRequestBuilder => {
    return createApiBuilderFromCtpClient(createCtClient()).withProjectKey({
        projectKey: process.env.CT_PROJECT_ID || '',
    });
};

export const getCustomerProfile = async (customerId: string): Promise<Customer> => {
    logger.info(`Getting customer ${customerId} in commercetools`);

    try {
        return (await getApiRoot().customers().withId({ ID: customerId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting customer in CT with id ${customerId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get customer api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }
};

export const getOrderById = async (orderId: string): Promise<Order | undefined> => {
    logger.info(`Getting order ${orderId} in commercetools`);
    let ctOrder: Order;

    try {
        ctOrder = (await getApiRoot().orders().withId({ ID: orderId }).get().execute()).body;
    } catch (error) {
        console.log(error);
        return undefined;
    }

    return ctOrder;
};

export const getOrderByPaymentId = async (paymentId: string): Promise<Order> => {
    logger.info(`Getting order with payment ${paymentId} in commercetools`);

    try {
        const orderResults = (await getApiRoot().orders().get({
            queryArgs: {
                limit: 1,
                where: `paymentInfo(payments(id = "${paymentId}"))`
            }
        }).execute()).body?.results;
        if (!orderResults?.length) {
            throw new StatusError(
                404,
                `No results returned when querying orders with payment ID ${paymentId}`,
            );
        }
        return orderResults[0];
    } catch (error: any) {
        logger.error(`Error getting order in CT with payment id ${paymentId}, status: ${error.status || error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get order by payment id api returns failed with status code ${error.status || error.statusCode}, msg: ${error.message}`,
        );
    }
};

export const getPaymentById = async (paymentId: string): Promise<Payment> => {
    logger.info(`Getting payment ${paymentId} in commercetools`);
    let ctPayment: Payment;

    try {
        ctPayment = (await getApiRoot().payments().withId({ ID: paymentId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting payment in CT with id ${paymentId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get payment api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }

    return ctPayment;
};
