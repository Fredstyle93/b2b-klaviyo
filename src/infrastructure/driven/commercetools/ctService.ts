import {
    AuthMiddlewareOptions,
    Client,
    ClientBuilder,
    QueueMiddlewareOptions,
    HttpMiddlewareOptions,
} from '@commercetools/sdk-client-v2';
import { ApiRoot, Payment, createApiBuilderFromCtpClient, Customer, Order } from '@commercetools/platform-sdk'
import logger from '../../../utils/log';
import fetch from 'node-fetch';
import { StatusError } from '../../../types/errors/StatusError';

type ByProjectKeyRequestBuilder = ReturnType<ApiRoot['withProjectKey']>;

let ctClient: Client;

export const getCtClient = (concurrency = 10): Client => {
    if (!ctClient) {
        const ctApiCredentials = JSON.parse(process.env.CT_API_CLIENT || '{"clientId":"","secret":""}');

        if(!process.env.CT_AUTH_URL){
            throw Error("Commercetools Host URL not set in environment variable CT_AUTH_URL. Set the variable and restart the application")
        }
        if(!process.env.CT_PROJECT_ID){
            throw Error("Commercetools Project ID not set in environment variable CT_PROJECT_ID. Set the variable and restart the application")
        }
        if(!process.env.CT_API_URL){
            throw Error("Commercetools API URL not set in environment variable CT_API_URL. Set the variable and restart the application")
        }
        const authMiddlewareOptions: AuthMiddlewareOptions = {
            host: process.env.CT_AUTH_URL,
            projectKey: process.env.CT_PROJECT_ID,
            credentials: {
                clientId: ctApiCredentials.clientId,
                clientSecret: ctApiCredentials.secret,
            },
            scopes: [process.env.CT_SCOPES || ''],
            fetch,
        };

        const httpMiddlewareOptions: HttpMiddlewareOptions = {
            maskSensitiveHeaderData: true,
            host: process.env.CT_API_URL,
            enableRetry: true,
            fetch,
        };

        const queueMiddlewareOptions: QueueMiddlewareOptions = {
            concurrency,
        };

        ctClient = new ClientBuilder()
            .withClientCredentialsFlow(authMiddlewareOptions)
            .withHttpMiddleware(httpMiddlewareOptions)
            .withQueueMiddleware(queueMiddlewareOptions)
            // .withLoggerMiddleware()
            .build();
    }
    return ctClient;
};

export const getApiRoot = (): ByProjectKeyRequestBuilder => {
    return createApiBuilderFromCtpClient(getCtClient()).withProjectKey({
        projectKey: process.env.CT_PROJECT_ID || '',
    });
};

export const getCustomerProfile = async (customerId: string): Promise<Customer> => {
    logger.info(`Getting customer ${customerId} in commercetools`);

    try {
        return (await getApiRoot().customers().withId({ ID: customerId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting customer in CT with id ${customerId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get customer api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }
};

//todo move to ctOrderService
export const getOrderById = async (orderId: string): Promise<Order | undefined> => {
    logger.info(`Getting order ${orderId} in commercetools`);
    let ctOrder: Order;

    try {
        ctOrder = (await getApiRoot().orders().withId({ ID: orderId }).get().execute()).body;
    } catch (error) {
        logger.error(error);
        return undefined;
    }

    return ctOrder;
};

//todo move to ctOrderService
export const getOrderByPaymentId = async (paymentId: string): Promise<Order> => {
    logger.info(`Getting order with payment ${paymentId} in commercetools`);

    try {
        const orderResults = (await getApiRoot().orders().get({
            queryArgs: {
                limit: 1,
                where: `paymentInfo(payments(id = "${paymentId}"))`,
                expand: 'paymentInfo.payments[*]',
            },
        }).execute()).body?.results;
        if (!orderResults?.length) {
            throw new StatusError(
                404,
                `No results returned when querying orders with payment ID ${paymentId}`,
            );
        }
        return orderResults[0];
    } catch (error: any) {
        logger.error(`Error getting order in CT with payment id ${paymentId}, status: ${error.status || error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get order by payment id api returns failed with status code ${error.status || error.statusCode}, msg: ${error.message}`,
        );
    }
};

//todo move to ctPaymentService
export const getPaymentById = async (paymentId: string): Promise<Payment> => {
    logger.info(`Getting payment ${paymentId} in commercetools`);
    let ctPayment: Payment;

    try {
        ctPayment = (await getApiRoot().payments().withId({ ID: paymentId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting payment in CT with id ${paymentId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get payment api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }

    return ctPayment;
};
