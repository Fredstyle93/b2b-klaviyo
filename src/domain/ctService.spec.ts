import { getCustomerProfile, getOrderById } from './ctService';
import { sampleCustomerApiResponse } from '../test/testData/ctCustomerApi';
import * as ctPlatformSdk from '@commercetools/platform-sdk';
import { sampleOrderCreatedMessage } from '../test/testData/orderData';

jest.mock('@commercetools/platform-sdk', () => {
    const module = jest.createMockFromModule<any>('@commercetools/platform-sdk');
    module.createApiBuilderFromCtpClient = jest.fn();
    return module;
});

describe('ctService > getCustomerProfile', () => {
    it('should get the customer profile from commercetools by id', async () => {
        jest.spyOn(ctPlatformSdk, 'createApiBuilderFromCtpClient').mockReturnValueOnce({
            withProjectKey: () => {
                return {
                    customers: () => {
                        return {
                            withId: () => {
                                return {
                                    get: () => {
                                        return {
                                            execute: () => {
                                                return {
                                                    body: sampleCustomerApiResponse,
                                                };
                                            },
                                        };
                                    },
                                };
                            },
                        };
                    },
                };
            },
        } as any);

        const customer = await getCustomerProfile('123456');

        expect(customer).toMatchSnapshot();
    });

    it('should throw error when the CT sdk throws error', async () => {
        jest.spyOn(ctPlatformSdk, 'createApiBuilderFromCtpClient').mockImplementationOnce(() => {
            throw Error('Unexpected Error');
        });

        await expect(getCustomerProfile('123456')).rejects.toThrow(Error);
    });
});

describe('ctService > getOrderById', () => {
    it('should get the order from commercetools by id', async () => {
        jest.spyOn(ctPlatformSdk, 'createApiBuilderFromCtpClient').mockReturnValueOnce({
            withProjectKey: () => {
                return {
                    orders: () => {
                        return {
                            withId: () => {
                                return {
                                    get: () => {
                                        return {
                                            execute: () => {
                                                return {
                                                    body: sampleOrderCreatedMessage.order,
                                                };
                                            },
                                        };
                                    },
                                };
                            },
                        };
                    },
                };
            },
        } as any);

        const order = await getOrderById('123456');

        expect(order).toMatchSnapshot();
    });

    it('should return undefined on error', async () => {
        jest.spyOn(ctPlatformSdk, 'createApiBuilderFromCtpClient').mockImplementationOnce(() => {
            throw Error('Unexpected Error');
        });

        const order = await getOrderById('123456');

        expect(order).toEqual(undefined);
    });
});
