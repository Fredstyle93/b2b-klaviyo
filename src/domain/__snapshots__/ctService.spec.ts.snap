// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ctService > getCustomerProfile should get the customer profile from commercetools by id 1`] = `
{
  "addresses": [
    {
      "city": "Brookefort",
      "country": "NL",
      "email": "jane.doe@example.com",
      "firstName": "Coleman",
      "id": "M5Q94Hg5",
      "lastName": "Pouros",
      "mobile": "+39 3234567892",
      "phone": "+39 3234567892",
      "postalCode": "12345",
      "streetName": "First Street",
      "streetNumber": "12",
      "title": "Mrs.",
    },
    {
      "city": "Example City",
      "country": "NL",
      "email": "jane.doe@example.com",
      "firstName": "Jane",
      "id": "VnDQjIJZ",
      "lastName": "Doe",
      "mobile": "+3112345679",
      "phone": "+3112345678",
      "postalCode": "12345",
      "streetName": "Third Street",
      "streetNumber": "34",
      "title": "Head of factory",
    },
  ],
  "authenticationMode": "Password",
  "billingAddressIds": [],
  "createdAt": "2023-02-01T14:14:56.605Z",
  "createdBy": {
    "clientId": "3bdbmCvX60kN6G-HfzLxcmlB",
  },
  "email": "Alfred_Lebsack@hotmail.com",
  "firstName": "Jailyn",
  "id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
  "isEmailVerified": false,
  "key": "testcustomer",
  "lastModifiedAt": "2023-02-01T14:14:56.605Z",
  "lastModifiedBy": {
    "clientId": "3bdbmCvX60kN6G-HfzLxcmlB",
  },
  "lastName": "Hilll",
  "password": "****bpU=",
  "shippingAddressIds": [],
  "stores": [],
  "title": "Miss",
  "version": 1,
}
`;

exports[`ctService > getOrderById should get the order from commercetools by id 1`] = `
{
  "createdAt": "2023-01-27T15:00:00.000Z",
  "customLineItems": [],
  "customerEmail": "test@klaviyo.com",
  "customerId": "123-123-123",
  "id": "3456789",
  "lastModifiedAt": "2023-01-27T15:00:00.000Z",
  "lineItems": [],
  "orderState": "Open",
  "origin": "Customer",
  "refusedGifts": [],
  "shipping": [],
  "shippingMode": "Single",
  "syncInfo": [],
  "totalPrice": {
    "centAmount": 1300,
    "currencyCode": "USD",
    "fractionDigits": 2,
    "type": "centPrecision",
  },
  "version": 24,
}
`;

exports[`ctService > getOrderByPaymentId should get the order from commercetools by payment id 1`] = `
{
  "createdAt": "2023-01-27T15:00:00.000Z",
  "customLineItems": [],
  "customerEmail": "test@klaviyo.com",
  "customerId": "123-123-123",
  "id": "3456789",
  "lastModifiedAt": "2023-01-27T15:00:00.000Z",
  "lineItems": [],
  "orderState": "Open",
  "origin": "",
  "paymentInfo": {
    "payments": [
      {
        "id": "3456789",
        "typeId": "payment",
      },
    ],
  },
  "refusedGifts": [],
  "returnInfo": [],
  "shipping": [],
  "shippingMode": "",
  "syncInfo": [],
  "totalPrice": {
    "centAmount": 1300,
    "currencyCode": "EUR",
    "fractionDigits": 2,
    "type": "centPrecision",
  },
  "version": 1,
}
`;

exports[`ctService > getPaymentById should get the payment from commercetools by id 1`] = `
{
  "amountPlanned": {
    "centAmount": 1000,
    "currencyCode": "USD",
    "fractionDigits": 2,
    "type": "centPrecision",
  },
  "createdAt": "2023-02-08T13:55:02.247Z",
  "id": "3456789",
  "interfaceId": "789011",
  "interfaceInteractions": [],
  "key": "123456",
  "lastModifiedAt": "2023-02-08T13:55:02.247Z",
  "paymentMethodInfo": {
    "method": "CREDIT_CARD",
    "name": {
      "en": "Credit Card",
    },
    "paymentInterface": "STRIPE",
  },
  "paymentStatus": {},
  "transactions": [
    {
      "amount": {
        "centAmount": 1000,
        "currencyCode": "USD",
        "fractionDigits": 2,
        "type": "centPrecision",
      },
      "id": "123456",
      "state": "Initial",
      "timestamp": "2015-10-20T08:54:24.000Z",
      "type": "Refund",
    },
  ],
  "version": 1,
}
`;
